kedro_vertexai/grouping.py

```diff
@@ -0,0 +1,116 @@
+from abc import ABC
+from dataclasses import dataclass, field
+from typing import Dict, Set, Tuple
+
+from kedro.pipeline.node import Node
+from toposort import CircularDependencyError, toposort
+
+TagsDict = Dict[str, Set[str]]
+
+
+@dataclass
+class Grouping:
+    nodes_mapping: Dict[str, Set[Node]] = field(default_factory=dict)
+    dependencies: Dict[str, Set[str]] = field(default_factory=dict)
+    # tags: TagsDict = field(default_factory=dict)
+
+    # not sure if this is good idea to hook it to initialization, but for our limited
+    # usage it should be fine
+    def __post_init__(self):
+        self.validate()
+
+    def validate(self):
+        try:
+            [_ for _ in toposort(self.dependencies)]
+        except CircularDependencyError as e:
+            raise GroupingException(
+                "Grouping has failed because of cyclic depedency after merging nodes. "
+                f"Check your group settings. {str(e)}"
+            )
+
+
+class GroupingException(Exception):
+    ...
+
+
+class NodeGrouper(ABC):
```

Comment:
I would add passing of both `run_config` as well as `kedro_context` to this class, someone will ask for this sooner or later. So the base class will accept `__init__(self, kedro_context: KedroContext, plugin_config: RunConfig)` and underlying classes will accept ` kedro_context: KedroContext, plugin_config: RunConfig` + whatever gets passed via `params:` in `vertexai.yml`.

Reply:
Not sure that I like implications of that. I can add and added draft of this change.

Why I don't like it:
- if the class gets all run config then its responsibility and expectations becomes to read its own params from config and interpret its structure and parse it somewhat...
- I would need to rethink/change class validation for Grouping config. It is not accessible to pass those arguments at that point and that stands somewhat in opposition to initialization of grouping class, which would be expected to use/read those arguments.

Check out the changes in last commit whether you like it or not.